site/src/lib/model/user.model.ts

```diff
@@ -66,72 +89,131 @@ export class User {
+        $and: [{ email }, { hasVerifiedEmail }],
+      });
+
+    return userDocument ? User.fromDocument(userDocument) : null;
+  }
+
+  static async getByShortname(
+    db: Db,
+    params: { shortname: string },
   ): Promise<User | null> {
-    const { email } = params;
+    const { shortname } = params;
 
     const userDocument = await db
       .collection<UserDocument>(User.COLLECTION_NAME)
-      .findOne({ email });
+      .findOne({ shortname });
 
     return userDocument ? User.fromDocument(userDocument) : null;
   }
 
   static async create(
     db: Db,
-    params: { preferredName: string; email: string },
+    params: {
+      email: string;
+      hasVerifiedEmail: boolean;
+      preferredName?: string;
+      shortname?: string;
+    },
   ): Promise<User> {
     const userProperties: UserProperties = {
       ...params,
       loginCodes: [],
+      emailVerificationCodes: [],
     };
 
     const { insertedId } = await db
       .collection<UserDocument>(User.COLLECTION_NAME)
       .insertOne(userProperties);
 
+    /** @todo: add to mailchimp mailing list */
+
     return new User({ id: insertedId.toString(), ...userProperties });
   }
 
-  async createLoginCode(db: Db): Promise<VerificationCode> {
-    const loginCode = await VerificationCode.create(db);
+  async update(
+    db: Db,
+    updatedProperties: Partial<UserProperties>,
+  ): Promise<User> {
+    if (this.shortname && updatedProperties.shortname) {
+      throw new Error("Cannot update shortname");
+    }
+
+    await db
+      .collection<UserDocument>(User.COLLECTION_NAME)
+      .updateOne({ _id: new ObjectId(this.id) }, { $set: updatedProperties });
+
+    merge(this, updatedProperties);
+
+    return this;
+  }
+
+  isSignedUp(): boolean {
+    return !!this.shortname && !!this.preferredName;
+  }
+
+  async createVerificationCode(
+    db: Db,
+    variant: VerificationCodeVariant,
+  ): Promise<VerificationCode> {
+    const verificationCode = await VerificationCode.create(db);
 
     await db.collection<UserDocument>(User.COLLECTION_NAME).updateOne(
       { _id: new ObjectId(this.id) },
       {
         $push: {
-          loginCodes: loginCode.toRef(),
+          [variant === "login" ? "loginCodes" : "emailVerificationCodes"]:
+            verificationCode.toRef(),
         },
       },
     );
 
-    return loginCode;
+    return verificationCode;
   }
 
-  async getLoginCode(
+  async getVerificationCode(
     db: Db,
-    params: { loginCodeId: string },
+    params: { verificationCodeId: string; variant: VerificationCodeVariant },
   ): Promise<VerificationCode | null> {
+    const { verificationCodeId, variant } = params;
+
+    const possibleVerificationCodes =
+      this[variant === "login" ? "loginCodes" : "emailVerificationCodes"];
+
     if (
-      !this.loginCodes.find(({ oid }) => oid.toString() === params.loginCodeId)
+      !possibleVerificationCodes.find(
+        ({ oid }) => oid.toString() === verificationCodeId,
+      )
```

Comment:
I think storing the `userId` and type on the codes rather than in the user->code direction would be better as it would allow you to get rid of this, in favour of trying to find a code in the db with the right id and type, but on balance I don't think it worth spending the time to update, since this is a temporary implementation.

Reply:
I'm honestly not sure why I didn't think to do this, have made this change in a recent commit becuase this would b bother me every time I see it (https://github.com/blockprotocol/blockprotocol/pull/83/commits/aeae7876d331e986a31acabbe971c6e9778a3064)